main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -2859,6 +2911,32 @@ object Weeder {
     case ParsedAst.Type.Ascribe(tpe, _, _) => leftMostSourcePosition(tpe)
   }
 
+  @tailrec
+  private def leftMostSourcePosition(eff: ParsedAst.Effect): SourcePosition = eff match {
```

Comment:
Could this be avoided?

Reply:
If we add more `SP`s to the parser, then likely yes. I'm not sure what the performance impact would be though.